src/QirRuntime/lib/QIR/callables.cpp

```diff
@@ -0,0 +1,189 @@
+struct QirTupleHeader
+{
+    int refCount = 0;
+};
+
+/*==============================================================================
+    Example of creating a callable
+
+    ; Single entry of a callable
+    ; (a callable might provide entries for body, controlled, adjoint and controlled-adjoint)
+    define void @UpdateAnsatz-body(
+        %TupleHeader* %capture-tuple, %TupleHeader* %arg-tuple, %TupleHeader* %result-tuple) { ... }
+
+    ; Definition of the callable
+    @UpdateAnsatz =
+        constant [4 x void (%TupleHeader*, %TupleHeader*, %TupleHeader*)*]
+        [
+            void (%TupleHeader*, %TupleHeader*, %TupleHeader*)* @UpdateAnsatz-body,
+            void (%TupleHeader*, %TupleHeader*, %TupleHeader*)* null,
+            void (%TupleHeader*, %TupleHeader*, %TupleHeader*)* null,
+            void (%TupleHeader*, %TupleHeader*, %TupleHeader*)* null
+        ]
+
+    %3 = call %Callable* @__quantum__rt__callable_create(
+        [4 x void (%TupleHeader*, %TupleHeader*, %TupleHeader*)*]* @UpdateAnsatz,
+        %TupleHeader* null)
+
+==============================================================================*/
+typedef void (*t_CallableEntry)(QirTupleHeader*, QirTupleHeader*, QirTupleHeader*);
+struct QirCallable
+{
+    static int constexpr Adjoint = 1;
+    static int constexpr Controlled = 1 << 1;
+
+  private:
+    static int constexpr TableSize = 4;
+
+    std::atomic<long> refCount;
+    t_CallableEntry function_table[QirCallable::TableSize];
+    QirTupleHeader* const capture;
+    int appliedFunctor = 0; // by default the callable is neither adjoint nor controlled
+
+    // prevent stack allocations
+    ~QirCallable()
+    {
+        assert(refCount == 0);
+    }
+
+  public:
+    QirCallable(const t_CallableEntry* ft_entries, QirTupleHeader* capture)
+        : refCount(1)
+        , capture(capture)
+        , appliedFunctor()
+    {
+        memcpy(this->function_table, ft_entries, QirCallable::TableSize * sizeof(void*));
+    }
+
+    QirCallable(const QirCallable* other)
+        : refCount(1)
+        , capture(other->capture)
+    {
+        memcpy(this->function_table, other->function_table, QirCallable::TableSize * sizeof(void*));
+    }
+
+    long AddRef()
+    {
+        int rc = ++this->refCount;
+        assert(rc != 1); // not allowed to resurrect!
+        return rc;
+    }
+
+    long Release()
+    {
+        assert(this->refCount > 0);
+
+        long rc = --this->refCount;
+        if (rc == 0)
+        {
+            delete this;
+        }
+        return rc;
+    }
+
+    void Invoke(QirTupleHeader* args, QirTupleHeader* result)
+    {
+        assert(this->appliedFunctor < QirCallable::TableSize);
+        this->function_table[this->appliedFunctor](capture, args, result);
+    }
+
+    void ApplyFunctor(int functor)
+    {
+        assert(functor == QirCallable::Adjoint || functor == QirCallable::Controlled);
+        this->appliedFunctor ^= functor;

```

Comment:
Is it intended that the functor is cleared if when trying to apply it is already present?

If it is, a comment would be great since my expectation would when applying would be to make sure that the functor bit is present regardless if it was already set before invoking this method.

If not, the more appropriate operator would be `|=`.

Reply:
The correct behavior, I think, is to clear the functor for Adjoint, but not for Controlled -- Adjoint Adjoint op is the same as op, but Controlled Controlled op is not the same as op, it's the same as Controlled op except that it requires some additional argument munging.

Reply:
Yes: A+A = I, A + C = C+ A = CA,  CA + A = C, but I haven't thought through the C + C case. Thank you for catching it!

Reply:
@alan-geller, I've fixed the `make_controlled` functor to behave as C + C = C but I don't think the runtime should be messing with the arguments as the tuple types are opaque to the runtime. Could you please take a look and either update the spec with what kind of argument munging is expected from the runtime or change the QIR generation code so I could enable the corresponding tests? Thanks!

Reply:
I'll specify the behavior in the documentation. Unfortunately the argument munging can't be avoided, but I think all that is necessary is to know the length of the tuple (and possibly not event that, I don't remember).

Reply:
But why the combining of the controls cannot happen during QIR generation? At the moment QIR creates the nested tuples, why cannot it flatten the controls rather than nesting them?